srml/staking/src/lib.rs

```diff
@@ -80,100 +175,250 @@ pub trait Trait: system::Trait + session::Trait {
+		/// 
+		/// See also `withdraw_unbonded`.
+		fn unbond(origin, #[compact] value: BalanceOf<T>) {
+			let controller = ensure_signed(origin)?;
+			let mut ledger = Self::ledger(&controller).ok_or("not a controller")?;
+
+			let mut value = value.min(ledger.active);
+
+			if !value.is_zero() {
+				ledger.active -= value;
+
+				// Avoid there being a dust balance left in the staking system.
+				let ed = T::Currency::minimum_balance();
+				if ledger.active < ed {
+					value += ledger.active;
+					ledger.active = Zero::zero();
+				}
 
-			// update nominators_for
-			t.swap_remove(target_index);
-			<NominatorsFor<T>>::insert(&target, t);
+				let era = Self::current_era() + Self::bonding_duration();
+				ledger.unlocking.push(UnlockChunk { value, era });
+				<Ledger<T>>::insert(&controller, ledger);
+			}
+		}
 
-			// update nominating
-			<Nominating<T>>::remove(&source);
+		/// Remove any unlocked chunks from the `unlocking` queue from our management.
+		/// 
+		/// This essentially frees up that balance to be used by the stash account to do
+		/// whatever it wants.
+		/// 
+		/// NOTE: This call must be made by the controller, not the stash.
+		/// 
+		/// See also `unbond`.
+		fn withdraw_unbonded(origin) {
+			let controller = ensure_signed(origin)?;
+			let ledger = Self::ledger(&controller).ok_or("not a controller")?;
+			<Ledger<T>>::insert(&controller, ledger.consolidate_unlocked(Self::current_era()));
+		}
 
-			// update bondage
-			<Bondage<T>>::insert(
-				source,
-				<system::Module<T>>::block_number() + Self::bonding_duration()
-			);
+		/// Declare the desire to validate for the origin controller.
+		///
+		/// Effects will be felt at the beginning of the next era.
+		/// 
+		/// NOTE: This call must be made by the controller, not the stash.
+		fn validate(origin, prefs: ValidatorPrefs<BalanceOf<T>>) {
+			let controller = ensure_signed(origin)?;
+			let mut _ledger = Self::ledger(&controller).ok_or("not a controller")?;
+			<Nominators<T>>::remove(&controller);
+			<Validators<T>>::insert(controller, prefs);
 		}
 
-		/// Set the given account's preference for slashing behaviour should they be a validator.
+		/// Declare the desire to nominate `targets` for the origin controller.
 		///
-		/// An error (no-op) if `Self::intentions()[intentions_index] != origin`.
-		fn register_preferences(
-			origin,
-			#[compact] intentions_index: u32,
-			prefs: ValidatorPrefs<BalanceOf<T>>
-		) {
-			let who = ensure_signed(origin)?;
-
-			if Self::intentions().get(intentions_index as usize) != Some(&who) {
-				return Err("Invalid index")
-			}
+		/// Effects will be felt at the beginning of the next era.
+		/// 
+		/// NOTE: This call must be made by the controller, not the stash.
+		fn nominate(origin, targets: Vec<<T::Lookup as StaticLookup>::Source>) {
+			let controller = ensure_signed(origin)?;
+			let mut _ledger = Self::ledger(&controller).ok_or("not a controller")?;
```

Comment:
Based on previous tests that did not allow a staked account to be a nominator and the impression that a `staked` account is somewhat _freezed_, shouldn't there be a test here that checks if `origin` is bonded or not?